modules/otr.cpp

```diff
@@ -0,0 +1,1291 @@
+                                          PROTOCOL_ID, &m_NewKey);
+        if (err) {
+            PutModuleBuffered(CString("Key generation failed: ") +
+                              gcry_strerror(err));
+            return;
+        }
+
+        PutModuleBuffered("Starting key generation in a background thread.");
+        m_GenKeyRunning = true;
+        AddJob(new COtrGenKeyJob(this));
+    }
+
+    void SaveIgnores() {
+        CString sFlat = "";
+        bool bFirst = true;
+
+        for (VCString::const_iterator it = m_vsIgnored.begin();
+             it != m_vsIgnored.end(); it++) {
+            if (bFirst) {
+                bFirst = false;
+            } else {
+                sFlat += " ";
+            }
+            sFlat += *it;
+        }
+        SetNV("ignore", sFlat, true);
+    }
+
+    bool IsIgnored(const CString& sNick) {
+        CString sNickLower = sNick.AsLower();
+
+        for (VCString::const_iterator it = m_vsIgnored.begin();
+             it != m_vsIgnored.end(); it++) {
+            if (sNickLower.WildCmp(*it)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    void CmdIgnore(const CString& sLine) {
+        if (sLine.Token(1).Equals("")) {
+            PutModuleBuffered("OTR is disabled for following nicks:");
+            for (VCString::const_iterator it = m_vsIgnored.begin();
+                 it != m_vsIgnored.end(); it++) {
+                PutModuleBuffered(*it);
+            }
+        } else if (sLine.Token(1).Equals("--remove")) {
+            CString sNick = sLine.Token(2);
+            if (sNick.Equals("")) {
+                PutModuleBuffered("Usage: ignore --remove nick");
+                return;
+            }
+
+            bool bFound = false;
+            for (VCString::iterator it = m_vsIgnored.begin();
+                 it != m_vsIgnored.end(); it++) {
+                if (it->Equals(sNick)) {
+                    m_vsIgnored.erase(it);
+                    bFound = true;
+                    break;
+                }
+            }
+
+            if (bFound) {
+                SaveIgnores();
+                PutModuleBuffered("Removed " + Clr(Bold, sNick) +
+                                  " from OTR ignore list.");
+            } else {
+                PutModuleBuffered("Not on OTR ignore list: " + sNick);
+            }
+        } else {
+            CString sNick = sLine.Token(1).MakeLower();
+            m_vsIgnored.push_back(sNick);
+            SaveIgnores();
+            PutModuleBuffered("Added " + Clr(Bold, sNick) +
+                              " to OTR ignore list.");
+        }
+    }
+
+    virtual bool OnLoad(const CString& sArgs, CString& sMessage) {
+        // Initialize libgcrypt for multithreaded usage
+        gcry_error_t err;
+        err = gcry_control(GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
+        if (err) {
+            sMessage = (CString("Failed to initialize gcrypt threading: ") +
```

Comment:
I think setting the callbacks multiple times to the same structure shouldn't be a problem. As you note below, problem would be if some other module or its dependency performed this operation with different argument.